test/test_linalg.py

```diff
@@ -4959,98 +4959,80 @@ def test_triangular_solve_out_errors_and_warnings(self, device, dtype):
-                            y = torch.arange(r * o * m * p, device=device).reshape(r, o, m, p)
-                            self.check_single_matmul(x, y, (r, o, n, p))
-
-                            # 2d, 4d, inner dimensions Fortran
-                            x = torch.arange(m * n, device=device).reshape(m, n).transpose(-1, -2)
-                            y = torch.arange(r * o * p * m, device=device).reshape(r, o, p, m).transpose(-1, -2)
-                            self.check_single_matmul(x, y, (r, o, n, p))
-
-                            # 2d, 4d, inner dimensions non-contiguous
-                            x = torch.arange(n * 2 * m, device=device).reshape(n, 2 * m)[:, ::2]
-                            y = torch.arange(r * o * m * 2 * p, device=device).reshape(r, o, m, 2 * p)[:, :, :, ::2]
-                            self.check_single_matmul(x, y, (r, o, n, p))
+        # test y.mT @ x.mT
+        xt = x.transpose(-2, -1) if x.ndim > 1 else x
+        yt = y.transpose(-2, -1) if y.ndim > 1 else y
+        expected = np.matmul(yt.cpu(), xt.cpu())
+
+        ans = torch.matmul(yt, xt)
+        self.assertTrue(ans.is_contiguous())
+        assertEqual(ans, expected)
+
+        out = torch.empty((0,), dtype=x.dtype, device=x.device)
+        ans = torch.matmul(yt, xt, out=out)
+        self.assertIs(ans, out)
+        self.assertTrue(ans.is_contiguous())
+        assertEqual(ans, expected)
+
+    def gen_sizes_matmul(self, x_dim, y_dim=4, matrix_size=8, batch_size=5):
+        """
+        Generates sequences of tuples (x, y) of with size(x) = x_dim and
+        size(y) <= y_dim that are compatible wrt. matmul
+        """
+        assert x_dim >= 1
+        assert y_dim >= 2
+        for by, mn in product(product(range(batch_size), repeat=y_dim - 2),
+                              product(range(matrix_size), repeat=2)):
+            if x_dim == 1:
+                size_x = mn[:1]
+                size_y = by + mn
+                yield size_x, size_y
+            else:
+                for bx, k in product(product(range(batch_size), repeat=x_dim - 2),
+                                     range(matrix_size)):
+                    size_x = bx + (k,) + mn[:1]
+                    size_y = by + mn
+                    yield size_x, size_y
+
+    @dtypesIfCUDA(torch.float, torch.complex64)  # Integer matmul just supported on CPU
+    @dtypes(torch.int64, torch.float, torch.complex64)
+    def test_matmul_small_brute_force_1d_Nd(self, device, dtype):
+        make_arg = partial(make_tensor, device=device, dtype=dtype)
+
+        for (size_x, size_y), nctg_x, nctg_y in product(self.gen_sizes_matmul(1), (True, False), (True, False)):
+            self.check_single_matmul(make_arg(size_x, noncontiguous=nctg_x),
+                                     make_arg(size_y, noncontiguous=nctg_y))
+
+    @dtypesIfCUDA(torch.float, torch.complex64)  # Integer matmul just supported on CPU
+    @dtypes(torch.int64, torch.float, torch.complex64)
```

Comment:
Does this noticeably increase test time? Looks like you are running more tests than previously. I'm not sure that those `ranges` are needed, maybe step of 2 will do if test time is a problem? They also don't test what #20452 was complaining about, which is 0-sized inputs.

Reply:
Zero sizes are tested as the ranges start at zero. I'll add also tests for the empty size.

The tests take about 40s in my box, but the tests before just tested longs on cpu, so this is expected. I think we could bruteforce sizes 0-4 and batches 0-3? Would that be a fair compromise?

Reply:
Oh right, I was looking at the previous test where ranges start at 1. Yeah, sizes 0-4, batches 0-3 should be enough.